containers/base/Dockerfile

```diff
@@ -15,98 +15,97 @@
+EXPOSE 8080
+
+# Path configuration
+ENV PATH $PATH:/tools/node/bin:/tools/google-cloud-sdk/bin
+ENV PYTHONPATH /env/python
+
 # Setup OS and core packages
 RUN echo "deb-src http://ftp.us.debian.org/debian testing main" >> /etc/apt/sources.list && \
     apt-get update -y && \
-    apt-get install --no-install-recommends -y -q \
-        curl wget unzip git vim build-essential ca-certificates pkg-config \
-        libatlas-base-dev liblapack-dev gfortran \
-        libpng-dev libfreetype6-dev libxft-dev \
-        rsync \
-        libxml2-dev \
-        python2.7 python-dev python-setuptools python-zmq openssh-client && \
+    apt-get install --no-install-recommends -y -q python unzip ca-certificates build-essential \
+    libatlas-base-dev liblapack-dev gfortran libpng-dev libfreetype6-dev libxft-dev libxml2-dev \
+    python2.7 python-dev python-setuptools openssh-client wget git pkg-config && \
     easy_install pip && \
-    mkdir -p /tools
+    mkdir -p /tools && \
 
 # Save GPL source packages
-RUN mkdir -p /srcs && \
+    mkdir -p /srcs && \
     cd /srcs && \
     apt-get source -d python-zmq wget git ca-certificates pkg-config libpng-dev && \
     wget https://mirrors.kernel.org/gnu/gcc/gcc-4.9.2/gcc-4.9.2.tar.bz2 && \
-    cd /
+    cd / && \
 
 # Setup Python packages. Rebuilding numpy/scipy is expensive so we move this early
 # to reduce the chance that prior steps can cause changes requiring a rebuild.
-RUN pip install -U numpy==1.11.0 && \
-    pip install -U pandas==0.18.0 && \
-    pip install -U scipy==0.17.0 && \
-    pip install -U scikit-learn==0.17.1 && \
-    pip install -U sympy==0.7.6.1 && \
-    pip install -U statsmodels==0.6.1
-
-RUN pip install -U tornado==4.3 \
-                   pyzmq==15.2.0 \
-                   jinja2==2.8 \
-                   jsonschema==2.5.1 \
-                   python-dateutil==2.5.0 \
-                   pytz==2015.4 \
-                   pandocfilters==1.3.0 \
-                   pygments==2.1.3 \
-                   argparse==1.2.1 \
-                   mock==1.2.0 \
-                   requests==2.9.1 \
-                   oauth2client==2.0.2 \
-                   httplib2==0.9.2 \
-                   futures==3.0.5 && \
-    pip install -U matplotlib==1.5.1 && \
-    pip install -U ggplot==0.6.8 && \
-    pip install -U seaborn==0.7.0 && \
-    pip install -U notebook==4.2.0 && \
-    pip install -U PyYAML==3.11 && \
-    pip install -U six==1.9.0 && \
-    pip install -U ipywidgets==5.2.2 && \
-    pip install -U future==0.15.2 && \
-    pip install -U psutil==4.3.0 && \
-    pip install -U google-api-python-client==1.5.1  && \
-    pip install -U plotly==1.12.5 && \
-    pip install -U nltk==3.2.1 && \
-    pip install -U bs4==0.0.1 && \
-    pip install -U crcmod==1.7 && \
-    pip install -U pillow==3.4.1 && \
-    easy_install pip && \
-    find /usr/local/lib/python2.7 -type d -name tests | xargs rm -rf
+    pip install -U --no-cache-dir numpy==1.11.0 && \
```

Comment:
"move them back to the end" --- does it have to be a different RUN group? As I recall, changing one item in a RUN group will invalidate all items in the same group. It's not just pydatalab, also tensorflow, cloudml SDK, where sometimes I experiment with different versions frequently. I just tested it and the build time starting from the first RUN group with numpy is about 7 minutes on my desktop. I hope that is not what we need to wait to test a cloudml SDK change.

No problem on the strategy of how we clean it up. Having to put everything into one RUN group is a surprise to me. So if we create two RUN groups, do we lose the space occupied by build files needed by one RUN group? Or do we lose the space for both? Maybe there is a better tradeoff (such as put TF, CloudML, PyDatalab together in a separate group, but I am not sure if it works).

That said, 7 minutes is not too bad if this is a docker limitation. I think we should favor run time performance over some development cost.